### PR TITLE
[Chore]: Technical: Isolate utils

### PR DESCRIPTION
**Pull request by dariaterekhovaae**

### CODE REVIEW

File: src/components/bottom-widget.tsx [hunk 1]
```diff
@@ -24,9 +24,9 @@ import TimeWidgetFactory from './filters/time-widget';
 import AnimationControlFactory from './common/animation-control/animation-control';
 import AnimationControllerFactory from './common/animation-control/animation-controller';
 import {ANIMATION_WINDOW, DIMENSIONS, FILTER_TYPES} from '@kepler.gl/constants';
-import {getIntervalBins} from 'utils/filter-utils';
+import {getIntervalBins} from '@kepler.gl/layers';
 import {media} from '@kepler.gl/styles';
-import {AnimationConfig, TimeRangeFilter} from 'reducers';
+import {AnimationConfig, TimeRangeFilter} from '@kepler.gl/types';
```
Comment (heshan0131) [inline src/components/bottom-widget.tsx:29]: add to `UPGRADE-GUIDE.md`
```md
### Moved from `kepler.gl/reducers` to `@kepler.gl/types`
- `AnimationConfig`
- `TimeRangeFilter`
```

File: src/components/common/animation-control/floating-time-display.tsx [hunk 2]
```diff
@@ -24,7 +24,7 @@ import classnames from 'classnames';
 import {Minus} from 'components/common/icons';
 import {DEFAULT_TIME_FORMAT} from '@kepler.gl/constants';
 import {CenterFlexbox} from 'components/common/styled-components';
-import {datetimeFormatter} from 'utils/data-utils';
+import {datetimeFormatter} from '@kepler.gl/layers';
```
Comment (heshan0131) [inline src/components/common/animation-control/floating-time-display.tsx:27]: why is this moved to `layers`? I don't see any layer using it. it is heavily used by components

File: src/components/common/animation-control/playback-controls.tsx [hunk 3]
```diff
@@ -26,7 +26,7 @@ import {Button, Tooltip, ButtonProps} from 'components/common/styled-components'
 import AnimationSpeedSliderFactory from './animation-speed-slider';
 import {Reset, Play, Pause, Rocket, AnchorWindow, FreeWindow} from 'components/common/icons';
 import {ANIMATION_WINDOW} from '@kepler.gl/constants';
-import {preciseRound} from 'utils/data-utils';
+import {preciseRound} from '@kepler.gl/layers';
```
Comment (heshan0131) [inline src/components/common/animation-control/playback-controls.tsx:29]: why is this moved to `layers`? I don't see any layer using it. it is heavily used by components

File: src/components/common/animation-control/speed-control.tsx [hunk 4]
```diff
@@ -23,7 +23,7 @@ import styled from 'styled-components';
 import AnimationSpeedSliderFactory from './animation-speed-slider';
 import {Button, CenterFlexbox} from 'components/common/styled-components';
 import {Rocket} from 'components/common/icons';
-import {preciseRound} from 'utils/data-utils';
+import {preciseRound} from '@kepler.gl/layers';
```
Comment (heshan0131) [inline src/components/common/animation-control/speed-control.tsx:26]: why is this moved to `layers`? I don't see any layer using it. it is heavily used by components

File: src/components/common/color-legend.tsx [hunk 5]
```diff
@@ -24,7 +24,7 @@ import {createSelector} from 'reselect';
 import {format} from 'd3-format';
 import moment from 'moment';
 import {SCALE_TYPES, SCALE_FUNC, ALL_FIELD_TYPES} from '@kepler.gl/constants';
-import {getTimeWidgetHintFormatter} from 'utils/filter-utils';
+import {getTimeWidgetHintFormatter} from '@kepler.gl/layers';
```
Comment (heshan0131) [inline src/components/common/color-legend.tsx:27]: why is this moved to `layers`? I don't see any layer using it. it is heavily used by components

File: src/components/common/data-table/cell-size.ts [hunk 6]
```diff
@@ -19,7 +19,7 @@
 // THE SOFTWARE.
 
 import document from 'global/document';
-import {parseFieldValue} from 'utils/data-utils';
+import {parseFieldValue} from '@kepler.gl/layers';
```
Comment (heshan0131) [inline src/components/common/data-table/cell-size.ts:22]: same as above

File: src/components/common/file-uploader/file-upload.tsx [hunk 7]
```diff
@@ -25,9 +25,9 @@ import UploadButton from './upload-button';
 import {DragNDrop, FileType} from 'components/common/icons';
 import FileUploadProgress from 'components/common/file-uploader/file-upload-progress';
 import FileDrop from './file-drop';
-import {FileLoading, FileLoadingProgress} from 'reducers/vis-state-updaters';
+import {FileLoading, FileLoadingProgress} from '@kepler.gl/types';
```
Comment (heshan0131) [inline src/components/common/file-uploader/file-upload.tsx:28]: add to `UPGRADE-GUIDE.md`

File: src/components/common/field-selector.tsx [hunk 8]
```diff
@@ -22,12 +22,13 @@ import React, {Component, ComponentType} from 'react';
 import styled from 'styled-components';
 import {createSelector} from 'reselect';
 
+import {Field} from '@kepler.gl/types';
+import {notNullorUndefined} from '@kepler.gl/layers';
```
Comment (heshan0131) [inline src/components/common/field-selector.tsx:26]: this is also weird, why layers

File: src/components/map-container.tsx [hunk 9]
```diff
@@ -43,35 +42,34 @@ import {
...
-  Datasets,
-  Filter,
-  InteractionConfig,
   MapControls,
-  MapState,
   MapStyle,
   Viewport
 } from 'reducers';
-import {SplitMapLayers} from 'reducers/vis-state-updaters';
+import { Datasets } from 'reducers/table-utils';
```
Comment (heshan0131) [inline src/components/map-container.tsx:72]: rebase with master should set lint to run on these files, which will catch this spacing issue here

File: src/components/modal-container.tsx [hunk 10]
```diff
@@ -26,8 +26,8 @@ import document from 'global/document';
 
 import {EXPORT_DATA_TYPE_OPTIONS, EXPORT_MAP_FORMATS} from '@kepler.gl/constants';
 import ModalDialogFactory from './modals/modal-dialog';
-import {exportJson, exportHtml, exportData, exportImage, exportMap} from 'utils/export-utils';
-import {isValidMapInfo} from 'utils/map-info-utils';
+import {isValidMapInfo} from '@kepler.gl/utils';
+import {exportJson, exportHtml, exportData, exportImage, exportMap} from '@kepler.gl/layers';
```
Comment (heshan0131) [inline src/components/modal-container.tsx:30]: these export functions shouldn't be in layers

File: src/components/plot-container.tsx [hunk 11]
```diff
@@ -24,16 +24,17 @@ import {createSelector} from 'reselect';
 import styled from 'styled-components';
 import {StaticMap} from 'react-map-gl';
 import debounce from 'lodash.debounce';
-import {exportImageError} from 'utils/notifications-utils';
+import {
+  exportImageError,
+  scaleMapStyleByResolution,
+  getCenterAndZoomFromBounds
+} from '@kepler.gl/utils';
+import {findMapBounds, convertToPng, getScaleFromImageSize} from '@kepler.gl/layers';
```
Comment (heshan0131) [inline src/components/plot-container.tsx:32]: `convertToPng` and `getScaleFromImageSize` shouldn't be in layers

File: src/components/bottom-widget.tsx [hunk 12]
```diff
@@ -24,9 +24,9 @@ import TimeWidgetFactory from './filters/time-widget';
 import AnimationControlFactory from './common/animation-control/animation-control';
 import AnimationControllerFactory from './common/animation-control/animation-controller';
 import {ANIMATION_WINDOW, DIMENSIONS, FILTER_TYPES} from '@kepler.gl/constants';
-import {getIntervalBins} from 'utils/filter-utils';
+import {getIntervalBins} from 'reducers/layer-utils/filter-utils';
```
Comment (heshan0131) [inline src/components/bottom-widget.tsx:27]: why is this in `layer-utils/filter-utils`? can't this just be under `reducers/filter-utils`?

File: src/components/common/animation-control/animation-control.tsx [hunk 13]
```diff
@@ -25,8 +25,7 @@ import classnames from 'classnames';
 import Slider from 'components/common/slider/slider';
 import {BottomWidgetInner} from 'components/common/styled-components';
 import PlaybackControlsFactory from './playback-controls';
-import FloatingTimeDisplayFactory from './floating-time-display';
-import {snapToMarks, datetimeFormatter} from 'utils/data-utils';
+import FloatingTimeDisplayFactory from './floating-time-display';import {datetimeFormatter, snapToMarks} from '@kepler.gl/utils';
```
Comment (heshan0131) [inline src/components/common/animation-control/animation-control.tsx:28]: new line missing

File: src/components/common/color-legend.tsx [hunk 14]
```diff
@@ -24,7 +24,7 @@ import {createSelector} from 'reselect';
 import {format} from 'd3-format';
 import moment from 'moment';
 import {SCALE_TYPES, SCALE_FUNC, ALL_FIELD_TYPES} from '@kepler.gl/constants';
-import {getTimeWidgetHintFormatter} from 'utils/filter-utils';
+import {getTimeWidgetHintFormatter} from 'reducers/layer-utils/filter-utils';
```
Comment (heshan0131) [inline src/components/common/color-legend.tsx:27]: same for this one. can this be in `reducers/layer-utils/filter-utils`?

File: src/components/common/field-selector.tsx [hunk 15]
```diff
@@ -22,12 +22,13 @@ import React, {Component, ComponentType} from 'react';
...
 import {createSelector} from 'reselect';
 
+import {Field} from '@kepler.gl/types';
+import {notNullorUndefined} from '@kepler.gl/utils';
+
 import ItemSelector from './item-selector/item-selector';
 import {classList} from './item-selector/dropdown-list';
-import {toArray} from 'utils/utils';
-import {notNullorUndefined} from 'utils/data-utils';
+import {toArray} from '@kepler.gl/utils';
```
Comment (heshan0131) [inline src/components/common/field-selector.tsx:30]: merge with the import on line 26

File: src/components/map-container.tsx [hunk 16]
```diff
@@ -45,33 +44,30 @@ import {
...
-  MapStyle,
-  Viewport
-} from 'reducers';
-import {SplitMapLayers} from 'reducers/vis-state-updaters';
+  getLayerHoverProp,
+  LayerHoverProp,
+  prepareLayersForDeck,
+  prepareLayersToRender,
+  renderDeckGlLayer
+} from 'reducers/layer-utils/layer-utils';
```
Comment (heshan0131) [inline src/components/map-container.tsx:70]: will you edit the path when isolation reducers? i feel like this should be either
'reducers/layer-utils/layer-utils` or `reducers/utils/layer-utils`

File: src/components/map/layer-hover-info.tsx [hunk 17]
```diff
@@ -23,8 +23,11 @@ import styled from 'styled-components';
 import {CenterFlexbox} from 'components/common/styled-components';
 import {Layers} from 'components/common/icons';
 import PropTypes from 'prop-types';
-import {notNullorUndefined} from 'utils/data-utils';
-import {getTooltipDisplayValue, getTooltipDisplayDeltaValue} from 'utils/interaction-utils';
+import {notNullorUndefined} from '@kepler.gl/utils';
+import {
+  getTooltipDisplayDeltaValue,
+  getTooltipDisplayValue
+} from 'reducers/layer-utils/interaction-utils';
```
Comment (heshan0131) [inline src/components/map/layer-hover-info.tsx:30]: Same here, this should be`reducers/interaction-utils`

File: src/components/modal-container.tsx [hunk 18]
```diff
@@ -26,8 +26,8 @@ import document from 'global/document';
 
 import {EXPORT_DATA_TYPE_OPTIONS, EXPORT_MAP_FORMATS} from '@kepler.gl/constants';
 import ModalDialogFactory from './modals/modal-dialog';
-import {exportJson, exportHtml, exportData, exportImage, exportMap} from 'utils/export-utils';
-import {isValidMapInfo} from 'utils/map-info-utils';
+import {exportHtml, isValidMapInfo, exportMap, exportJson} from '@kepler.gl/utils';
+import {exportData} from 'reducers/layer-utils/export-utils';
```
Comment (heshan0131) [inline src/components/modal-container.tsx:30]: `exportData` is not used by any reducer function, it's only used by component, can we move it to `@kepler.gl/utils`?

File: src/reducers/table-utils/kepler-table.ts [hunk 19]
```diff
@@ -19,50 +19,36 @@
...
 import {ascending, descending} from 'd3-array';
 
 // import {validateInputData} from 'processors/data-processor';
-import {generateHashId} from 'utils/utils';
-import {getGpuFilterProps, getDatasetFieldIndexForFilter} from 'utils/gpu-filter-utils';
+import {TRIP_POINT_FIELDS, SORT_ORDER, ALL_FIELD_TYPES, SCALE_TYPES} from '@kepler.gl/constants';
+import {RGBColor, Field, FieldPair, FieldDomain, Filter, ProtoDataset} from '@kepler.gl/types';
+
+import {generateHashId, getSortingFunction} from '@kepler.gl/utils';
+import {getGpuFilterProps, getDatasetFieldIndexForFilter} from '../layer-utils/gpu-filter-utils';
```
Comment (heshan0131) [inline src/reducers/table-utils/kepler-table.ts:29]: `getGpuFilterProps, getDatasetFieldIndexForFilter` should be moved to `table-utils`

**Review by heshan0131 — COMMENTED** [review]
I realize i have been making same comment in multiple places, so i am just going to comment it here

I see a lot of `import xx from reducers/layer-utils/xx-utils` I don't understand why they got moved into `reducers/layer-utils` path. if you need those utils in reducers module, can you move them out of `layer-utils`? they are not used by layers. change `reducers/layer-utils/xx-utils` to `reducers/xx-utils`

File: src/utils/src/dataset-utils.ts [hunk 20]
```diff
@@ -0,0 +1,539 @@
+// Copyright (c) 2022 Uber Technologies, Inc.
```
Comment (heshan0131) [inline src/utils/src/dataset-utils.ts:1]: this file is created as new...instead of moved. Can we revert it?

File: test/browser/layer-tests/geojson-layer-specs.js [hunk 21]
```diff
@@ -19,8 +19,10 @@
 // THE SOFTWARE.
 
 import test from 'tape';
-import {KeplerGlLayers, defaultElevation, defaultLineWidth, defaultRadius} from '@kepler.gl/layers';
+import {defaultElevation, defaultLineWidth, defaultRadius} from '@kepler.gl/layers';
+import {copyTableAndUpdate} from 'reducers/table-utils';
 
+import {KeplerGlLayers} from '@kepler.gl/layers';
```
Comment (heshan0131) [inline test/browser/layer-tests/geojson-layer-specs.js:25]: merge 2 `'@kepler.gl/layers';` imports

File: test/browser/layer-tests/grid-layer-specs.js [hunk 22]
```diff
@@ -31,7 +31,8 @@ import {
   pointLayerMeta
 } from 'test/helpers/layer-utils';
 
-import {KeplerGlLayers, pointToPolygonGeo} from '@kepler.gl/layers';
+import {pointToPolygonGeo} from '@kepler.gl/layers';
+import {KeplerGlLayers} from '@kepler.gl/layers';
```
Comment (heshan0131) [inline test/browser/layer-tests/grid-layer-specs.js:35]: merge these 2 imports...

**Comment by heshan0131** [comment]
Approving it now. please make sure to address this comment 
https://github.com/keplergl/kepler.gl/pull/1876/files#r940761888

**Review by heshan0131 — APPROVED** [review]

File: src/utils/src/dataset-utils.ts (hunk 20 above)
Comment (dariaterekhovaae) [reply src/utils/src/dataset-utils.ts:1]: Unfortunately git recognized moved files itself, there's no way to manually mark a file as moved. And it looks like due to a big difference with the original file this one isn't recognized so... I tried to play with a commits a bit but it looks that git only estimates difference between the start commit and the end one, so there's no way I can make it think that file is moved with the current changes...

Comment (heshan0131) [reply src/utils/src/dataset-utils.ts:1]: that's fine, let's land this
